app/models/api/openid_connect/o_auth_application.rb

```diff
@@ -0,0 +1,97 @@
+require "digest"
+
+module Api
+  module OpenidConnect
+    class OAuthApplication < ActiveRecord::Base
+      has_many :authorizations, dependent: :destroy
+      has_many :user, through: :authorizations
+
+      validates :client_id, presence: true, uniqueness: true
+      validates :client_secret, presence: true
+      validates :client_name, uniqueness: {scope: :redirect_uris}
+
+      %i(redirect_uris response_types grant_types contacts jwks).each do |serializable|
+        serialize serializable, JSON
+      end
+
+      before_validation :setup, on: :create
+      before_validation do
+        redirect_uris.sort!
+      end
+
+      def setup
+        self.client_id = SecureRandom.hex(16)
+        self.client_secret = SecureRandom.hex(32)
+      end
+
+      def image_uri
+        logo_uri ? Diaspora::Camo.image_url(logo_uri) : nil
+      end
+
+      class << self
+        def available_response_types
+          ["id_token", "id_token token", "code"]
```

Comment:
@AugierLe42e %w(id_token id_token token code) is not the same as ["id_token", "id_token token", "code"]